[FEATURE] Add metadata to QGIS projects

nyalldawson opened the pull request:
Implements the first part of https://github.com/qgis/QGIS-Enhancement-Proposals/issues/115

Allows for detailed metadata to be set in project properties, including abstracts, keywords, contact details, links, etc.

Developed for Arpa Piemonte (Dipartimento Tematico Geologia e Dissesto) within ERIKUS project

TODO: 
~~- expression context variables for project metadata items~~
~~- include metadata in layout exports~~

nyalldawson committed:
Split metadata into two classes

A new base class QgsMetadataBase has been split off QgsLayerMetadata,
containing all non-layer specific metadata items.

This allows us to reuse the common metadata settings for other
non-layer items, e.g. projects.

nyalldawson committed:
Create unit test for QgsMetadataBase

nyalldawson committed:
Update metadata validator for base class split

nyalldawson committed:
Add new QgsProjectMetadata metadata class

For storing QgsProject metadata. Contains all metadata items
from QgsMetadataBase, plus adds author and creation datetime
metadata items.

nyalldawson committed:
Add QgsProjectMetadata member to QgsProject

nyalldawson committed:
Make project title getters/setters a shortcut to metadata title field

And silently upgrade old project titles to metadata titles

nyalldawson committed:
Code shuffle to avoid duplicate code

nyalldawson committed:
Update metadata xml definitions

nyalldawson committed:
Add project metadata mode to QgsMetadataWidget

nyalldawson committed:
[FEATURE] Add metadata to QGIS projects

Allows for detailed metadata to be set in project properties,
including abstracts, keywords, contact details, links, etc.

Developed for Arpa Piemonte (Dipartimento Tematico Geologia e Dissesto)
within ERIKUS project

nirvn commented:
@nyalldawson , I'm wondering whether it'd be better to have a function to access project (and eventually layer) metadata instead of exposing those through variables, e.g. metadata_project('author'). 

That would stand quite nicely alongside a possible metadata_layer('layer_name','metadata_name') function.

nirvn commented:
Just to be clear, there's indeed nothing wrong with making project metadata available through variables, but that ultimately wouldn't work for layers, in the context of accessing such information in e.g. a expression-based layout label item.

nyalldawson committed:
Spelling

nyalldawson committed:
Fix accidental class rename

nyalldawson committed:
Fix sip subclassing of metadata subclasses

nyalldawson committed:
Add tests for QgsMetadataWidget

nyalldawson committed:
Synchronize project title field and project metadata title
field in project properties dialog

Since these two are equivalent, we automatically keep them
in sync.

nyalldawson committed:
Update more labels to reflect project metadata in widget

nyalldawson committed:
Standardise capitalisation

nyalldawson committed:
When a new project is created, default to setting the metadata
author to the current logged on user name

nyalldawson committed:
[FEATURE] Add expression context variables for project metadata

- @project_author
- @project_abstract
- @project_creation_date
- @project_identifier
- @project_keywords

Allows retrieval of project metadata through QGIS expressions

Developed for Arpa Piemonte (Dipartimento Tematico Geologia e Dissesto)
within ERIKUS project

nyalldawson commented:
@nirvn 

>  I'm wondering whether it'd be better to have a function to access project (and eventually layer) metadata instead of exposing those through variables, e.g. metadata_project('author').

> That would stand quite nicely alongside a possible metadata_layer('layer_name','metadata_name') function.

Probably, yes. But I'd still opt for some explicit variables for simplicity also (as is done in this PR) 

It would be very nice to be able to do something like:

    array_to_string(apply(@project_layers,metadata_layer(@element,'attribution'),', ')

(where @project_layers = array of layer ids for layers in project and apply = function to apply a sub-expression to all elements in an array (where @element is the current array element)

nyalldawson committed:
[FEATURE][layouts] Export project metadata as SVG RDF metadata

Adds an option to include project metadata into SVG exports
generated from layouts, using the SVG RDF standard.

Developed for Arpa Piemonte (Dipartimento Tematico Geologia e Dissesto)
within ERIKUS project

nirvn commented:
@nyalldawson , nice. While we're dreaming about what's possible, I'd add a is_layer_visible call there:

`array_to_string(apply(@project_layers,if(is_layer_visible(@element),metadata_layer(@element,'attribution'), null),',')`


Nice :)

nyalldawson commented:
@nirvn

Actually we'd probably want to get the list of layers from a layout map item directly...

nirvn commented:
@nyalldawson , right. I was thinking it'd be nice for the text decoration to support insertion of expressions (the same way layout label items do ATM). We could do all sorts of cool things there, inc. having attribution text based on visibility of layer(s) within canvas extent, etc.

nirvn commented on src/core/qgsproject.cpp:
```diff
@@ -511,6 +511,7 @@ void QgsProject::clear()
   mCustomVariables.clear();
   mMetadata = QgsProjectMetadata();
   mMetadata.setCreationDateTime( QDateTime::currentDateTime() );
+  mMetadata.setAuthor( QgsApplication::userFullName() );
```
I'm -0 on this -- working in a sensitive environment, and working with other people in the same environment, I cannot recall the number of times someone had his/her name (or a user name that's easily linked to the person) attached to a word document or PDF without his/her knowledge.

I recon this isn't necessarily the most common scenario, just serving food for thoughts :)

nyalldawson replied:
How about a hidden settings key which avoids this?

nirvn replied:
@nyalldawson , sounds like a reasonable compromise.

nyalldawson replied:
Done

nyalldawson committed:
Fix unit test

NathanW2 commented:
Hidden option to turn on or off?

On Tue, 20 Mar. 2018, 2:10 pm Mathieu Pellerin, <notifications@github.com>
wrote:

> *@nirvn* commented on this pull request.
> ------------------------------
>
> In src/core/qgsproject.cpp
> <https://github.com/qgis/QGIS/pull/6635#discussion_r175651224>:
>
> > @@ -511,6 +511,7 @@ void QgsProject::clear()
>    mCustomVariables.clear();
>    mMetadata = QgsProjectMetadata();
>    mMetadata.setCreationDateTime( QDateTime::currentDateTime() );
> +  mMetadata.setAuthor( QgsApplication::userFullName() );
>
> @nyalldawson <https://github.com/nyalldawson> , sounds like a reasonable
> compromise.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/6635#discussion_r175651224>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3Olv8HI6FrgeLuW1CzsjBcvNpKSZks5tgIEWgaJpZM4SvtkF>
> .
>

nyalldawson commented:
> Hidden option to turn on or off?

Turn off I say

nyalldawson committed:
[FEATURE][layouts] Export project metadata in PDF/image exports

Includes project metadata in PDF exports, and supported image
based formats.

Image based metadata support depends on the format and the
Qt library's handling of each particular format (e.g. PNG
outputs are well supported).

Developed for Arpa Piemonte (Dipartimento Tematico Geologia e Dissesto)
within ERIKUS project

nyalldawson committed:
Add settings key to anonymize new projects

Defaults to false, but if set will not automatically set
metadata for newly created projects such as author
and creation date.

nirvn commented:
@nyalldawson , thanks for incorporating this anonymity option.

nirvn commented:
@nyalldawson , just tested the PR, nice work.

Few thoughts:
- The metadata tab in the project properties window could be better placed lower, to adopt the location found in the vector and raster properties windows
- I noticed that the current user name is added to author metadata when opening a pre-existing project file, I'm wondering whether the author file should be left untouched here
- Do we need a "type" field for project? If so, the default values listed in the drop down combo box are a bit out of context

Also, and that's not specific to this PR at all (and should not block this), it's a bit weird to have the contacts panel and links panel offering a totally different UX to enter items. I'd favor the contacts panel method (i.e. showing line entries for each parameter of a contact entry) is we decide to harmonize that.

nyalldawson commented:
> The metadata tab in the project properties window could be better placed lower, to adopt the location found in the vector and raster properties windows

I actually disagree here - I'd rather see the layer one moved higher. In my opinion metadata is one of the "fundamental" aspects of a project/layer, so accordingly should be higher placed.

nyalldawson commented:
> I noticed that the current user name is added to author metadata when opening a pre-existing project file, I'm wondering whether the author file should be left untouched here

nyalldawson committed:
Remove autogenerated metadata when loading an older project

nyalldawson commented:
> Do we need a "type" field for project? If so, the default values listed in the drop down combo box are a bit out of context

Agreed - it's freeform, so I'll leave it -- but none of the existing ISO standard options really fit. "collectionSession" or "series" is closest... but I think we just leave it empty by default until some more fitting codes are added to the standard. https://geo-ide.noaa.gov/wiki/index.php?title=ISO_19115_and_19115-2_CodeList_Dictionaries#MD_ScopeCode describes some extra ones, where "product" or "document" could be good candidates, but we better wait till they are added to the standard first.

nyalldawson commented:
> it's a bit weird to have the contacts panel and links panel offering a totally different UX to enter items. I'd favor the contacts panel method (i.e. showing line entries for each parameter of a contact entry) is we decide to harmonize that.

I don't really understand this comment - they do both use the list-based approach. It's just that the contact tab has the extra (singular) fields shown at the top

nirvn commented:
> I actually disagree here - I'd rather see the layer one moved higher. In my opinion metadata is one of the "fundamental" aspects of a project/layer, so accordingly should be higher placed.

I'd be fine with that, makes sense.

> I don't really understand this comment - they do both use the list-based approach. It's just that the contact tab has the extra (singular) fields shown at the top

God, forget this, I was confused. All good.

nirvn commented:
@nyalldawson , just tested the updated PR, IMHO ready to merge :tada:

nyalldawson commented:
Thanks @nirvn!
